perfkitbenchmarker/providers/ibmcloud/ibmcloud_virtual_machine.py

```diff
@@ -0,0 +1,599 @@
+    self.key = cmd.ListResources()
+    logging.info('Key found: %s', self.key)
+    if self.key is None:
+      cmd.flags['items'] = 'keys'
+      cmd.flags['prefix'] = self.prefix + FLAGS.run_uri
+      self.key = self._CreateRiasKey()
+      if self.key is None:
+        raise errors.Error('IBM Cloud ERROR: Failed to create a rias key')
+      logging.info('Created a new key: %s', self.key)
+
+    logging.info('Looking up the image: %s', self.imageid)
+    cmd.flags['imageid'] = self.imageid
+    self.os_data = util.GetOsInfo(cmd.ImageShow())
+    logging.info('Image os: %s', self.os_data)
+    logging.info('Checking mzone setup finished')
+
+  def _DeleteKey(self):
+    """Deletes the rias key."""
+    with self._lock:
+      # key is not dependent on vpc, one key is used
+      if self.key not in IbmCloudVirtualMachine.validated_resources_set:
+        time.sleep(5)
+        cmd = ibm.IbmAPICommand(self)
+        cmd.flags['items'] = 'keys'
+        cmd.flags['id'] = self.key
+        cmd.DeleteResource()
+        IbmCloudVirtualMachine.validated_resources_set.add(self.key)
+
+  def _Create(self):
+    """Creates and starts a IBM Cloud VM instance."""
+    self._CreateInstance()
+    if self.subnet:  # this is for the primary vnic and fip
+      self.fip_address, self.fip_id = self._CreateFip(self.name + 'fip')
+      self.ip_address = self.fip_address
+      self.internal_ip = self._WaitForIPAssignment(self.subnet)
+      logging.info('Fip: %s, ip: %s', self.ip_address, self.internal_ip)
+
+    if self.subnets:
+      # create the extra vnics
+      cmd = ibm.IbmAPICommand(self)
+      cmd.flags['instanceid'] = self.vmid
+      for subnet_name in self.subnets.keys():
+        cmd.flags['name'] = subnet_name
+        cmd.flags['subnet'] = self.subnets[subnet_name]['id']  # subnet id
+        logging.info('Creating extra vnic for vmid: %s, subnet: %s', \
+                     self.vmid, cmd.flags['subnet'])
+        vnicid, ip_addr = cmd.InstanceVnicCreate()
+        logging.info('Extra vnic created for vmid: %s, vnicid: %s, ip_addr: %s',
+                     self.vmid, vnicid, ip_addr)
+        self.subnets[subnet_name]['vnicid'] = vnicid
+        self.subnets[subnet_name]['ip_addr'] = ip_addr
+      logging.info('Extra vnics created for vmid: %s, subnets: %s',
+                   self.vmid, self.subnets)
+
+  def _Delete(self):
+    """Delete all the resources that were created """
+    if self.vm_deleted:
+      return
+    if not ibm.IbmAPICommand.ibmcloud_auth_token or \
+    time.time() - ibm.IbmAPICommand.ibmcloud_auth_token_time > FLAGS.ibmcloud_timeout:
+      self._CheckLogin()
+    self._StopInstance()
+    if self.fip_address:
+      self._DeleteFip()
+    time.sleep(10)
+    self._DeleteInstance()
+    if not FLAGS.ibmcloud_resources_keep:
+      self.network.Delete()
+    self._DeleteKey()
+
+  def _DeleteDependencies(self):
+    """Delete dependencies."""
+    pass
+
+  def _Exists(self):
+    if self.vm_deleted:
```

Comment:
This doesn't seem right -- wouldn't it be on initialization of this class that _Exists() will return True and it hasn't been delete nor has the instance_stop_failed?

Reply:
yes that instance_stop_failed was not being used correctly, it's actually not really needed, removed unnecessary flags.